### ISSUE TITLE
"return null" within "for loop" within async method does not respect strictNullChecks

### ISSUE THREAD

**Issue by willportnoy** [Microsoft/TypeScript#14298]
**TypeScript Version:**  2.2.1

strictNullChecks set to true

**Code**

```ts
type Func_Async<R> = () => Promise<R>;

// no errors here, expected error like below examples
const fail_NoNullError: Func_Async<Object> =
   async () => {
        for (let index = 0; index < 1; ++index) {
            return null;
        }

        return {};
    };

// as expected
// error TS2322: Type '() => Promise<null>' is not assignable to type 'Func_Async<Object>
const good_HasNullError: Func_Async<Object> =
   async () => {
        return null;
    };

type Func_Sync<R> = () => R;

// as expected:
// error TS2322: Type '() => {} | null' is not assignable to type 'Func_Sync<Object>'.
const good_HasNullErrorSync: Func_Sync<Object> =
    () => {
        for (let index = 0; index < 1; ++index) {
            return null;
        }

        return {};
    };

```

**Expected behavior:**

I expected "return null" in the middle of a for loop within an async method should produce a compiler error with strictNullChecks, much like the other sync and async examples above.

**Actual behavior:**

The code compiles without any errors.

**Comment by aluanhaddad** [comment]
I think the issue may stem from the type signature of .then

**Comment by mhegazy** [comment]
This is stems from the same issue detailed in https://github.com/Microsoft/TypeScript/issues/13970
